COSMA's FindMKL and FindMKL_ScaLAPACK

teonnik opened the pull request:
MKL modules used in COSMA

- Support for sequential, Intel OpenMP and GNU OpenMP back ends
- Support for 32bit and 64bit interfaces
- Support for ScaLAPACK with both MPICH and OpenMPI backends

TODO:

- [x]  Rename `COSMA...` variables

teonnik committed:
Add FindMKL and FindMKL_ScaLAPACK

- Support for sequential, Intel OpenMP and GNU OpenMP back ends
- Support for 32bit and 64bit interfaces
- Support for ScaLAPACK with both MPICH and OpenMPI backends

albestro commented on FindMKL.cmake:
```diff
@@ -0,0 +1,110 @@
+
+if(NOT DEFINED MKL_64BIT)
+    set(MKL_64BIT OFF)
+endif()
```
I suggest to use `CACHE` variables for options, so the user can change them easily with tools like `ccmake`.

teonnik replied:
I am not a huge fan of `CACHE` variables to be honest. They are global and complicate bundling of Find modules when installing a great deal.

That was my experience from integrating COSMA into SIRIUS. There were clashes between the FindMKL module provided by SIRIUS and the one bundled with COSMA. The clashes were due to similarly named `CACHE` variables.

teonnik replied:
Having said that, a lot of variables in the `FindMKL` script are already `CACHE` variables, e.g. all the variables from `find_path` and `find_library`. Perhaps not having `MKL_ROOT`, `MKL_PARALLEL` and `MKL_64BIT` as `CACHE` doesn't really help that much.

albestro replied:
I agree with you that having multiple find-modules for the same package leads to problems. 
This is another good reason to agree on a single script and use it everywhere! 😁

What does CMake Gods think of **CMake CACHE variables**?

@halfflat @havogt @lukasm91 @biddisco

halfflat replied:
As a general rule, I loathe them for their surprising statefulness, but I try to follow `find_package` conventions regarding their use, and use them for user-visible configuration options.

havogt replied:
If they are not cached, how do you use them currently? They are meant to be configured by the caller of `find_package`?

albestro commented on FindMKL.cmake:
```diff
@@ -0,0 +1,110 @@
+                      lib
+        )
+    mark_as_advanced(${_name})
+endfunction()
```
As coding style suggestion, usually I write functions before the "main" code (I'm using this coding style after looking at different official find-modules).

albestro commented on FindMKL.cmake:
```diff
@@ -0,0 +1,110 @@
+__mkl_find_library(MKL_CORE_LIB mkl_core)
+
+set(_mkl_lp "lp64")
+mark_as_advanced(${_mkl_lp})
```
`mark_as_advanced` is just for `CACHE` variables, and `_mkl_lp` is not, so it is not useful.

albestro commented on check_for_openmpi.cmake:
```diff
@@ -0,0 +1,11 @@
+  else()
+    set(COSMA_WITH_OPENMPI ON PARENT_SCOPE)
+  endif()
+endfunction()
```
I think it would be better to change this script to detect the type of MPI implementation found by [FindMPI](https://cmake.org/cmake/help/v3.15/module/FindMPI.html), by introspection on variables.

It seems that [MPI_DETERMINE_LIBRARY_VERSION](https://cmake.org/cmake/help/v3.15/module/FindMPI.html#advanced-variables-for-using-mpi) may be helpful for this; if it (more or less) clearly states the implementation, I would go for embedding this detection directly into the MKL_ScaLAPACK section without having an additional cmake script for this function.

albestro commented on FindMKL_ScaLAPACK.cmake:
```diff
@@ -0,0 +1,62 @@
+include(FindPackageHandleStandardArgs)
+
+find_package(MPI REQUIRED)
+find_package(MKL REQUIRED)
```
As already said in the main comment, I suggest to merge it in the main MKL script and enable it as a COMPONENT like in #2.

albestro reviewed (commented):
Just a few general comments.

In general, I think that the management of the library has a better structure than in #2: it addresses more variants and use the `find_library` to detect them.

For what concerns modules like ScaLAPACK, I would go for merging them in the same script as COMPONENTs, since they are not a separate thing.

teonnik committed:
Merge FindMKL and FindMKL_ScaLAPACK

- Implement ScaLAPACK as a component of FindMKL.
- Improve FindMKL documentation
- Move function declarations to the top
- Use CACHE variables for options but guard them with `NOT DEFINED`
- Rename OPENMPI variables to be project agnostic

teonnik commented:
This is ready for another round of reviews. @albestro  @halfflat @biddisco @havogt @lukasm91

albestro commented on FindMKL.cmake:
```diff
@@ -0,0 +1,227 @@
+#   `set(... CACHE ...)` is still stuck with the old behaviour. 
+#
+#   https://cmake.org/cmake/help/v3.15/command/set.html#set-cache-entry
+#   https://cmake.org/cmake/help/v3.15/command/set.html#set-cache-entry
```
Repeated line.

albestro commented on FindMKL.cmake:
```diff
@@ -0,0 +1,227 @@
+    __mkl_find_library(MKL_THREADING_LIB mkl_sequential)
+endif ()
+
+# Find ScaLAPACK
```
I propose to open a PR on this branch with as objective the management of components.

I've just opened #3 for keeping track of this.

teonnik replied:
Is there something wrong with how I manage components here?

Currently the module searches for core MKL, i.e. BLAS/LAPACK. The ScaLAPACK component is searched for and made available if found even if the component is not explicitly specified. The component is handled correctly when explicitly specified and non-optional. I feel like having BLAS/LAPACK as a component is strange as it is the core of MKL and without it, none of the other components (i.e. ScaLAPACK or FFT (not yet implemented)) will work anyway.

albestro replied:
See #3, It would be better to follow the philosophy "Pay Only For What You Use".

For what concerns your PR, you may just address ScaLAPACK as an optional component. I would accept also not separating it from BLACS for the moment, waiting for #3 to list all components and then managing them there with all dependencies.

albestro commented on FindMKL.cmake:
```diff
@@ -0,0 +1,227 @@
+# Options
+#
+# The `NOT DEFINED` guards on CACHED variables are needed to make sure that 
+# normal variables with of same name always take precedence*.
```
This is an interesting topic.

As far as I know, `if` does not check just normal variables, but also CACHE variables. So, once a cache variable with the same name is set, the normal one will not have priority over the cached one.

This has some minor edge cases that leads to problems in case you are working on the CMake side, however, I think that it may be a candidate solution for the problem in https://github.com/eth-cscs/cmake-recipes/pull/2#issuecomment-537828201

@halfflat @havogt what do you think about this solution?

halfflat replied:
To be honest, I'm not sure I understand the desired semantics. IIRC, this code will:
* If the variable is in cache, don't do anything (e.g. from a previous run, or from a user-supplied `-Dvar=...`).
* If the variable is defined but not in cache, don't change, but also don't make a cache variable — this would be where at the calling site, something in the CMakeLists.txt has set it.
* If the variable is not defined at all, then make a cache variable with a reasonable default.

Is this correct?

albestro replied:
I think that your summary is complete and correct, and you are pointing out a few downsides of this solution that the one in #2 seems to solve.

I want to be able to set values from the script, so that in my package config script, when I do the `find_dependency`, I can force the FindMKL package to replicate exactly the same options as configured at build time. With this solution, values set from the script will be not considered if the user sets a cache variable with `-D` from the command line (as you are pointing out in your first point).

Moreover, with the alternative solution in #2 of using "*_CUSTOM" variables as prioritized, related CACHE variables are always set to show to the user the final value that will be used. Anyway, this may still be a bit misleading because it discards any change made by the user to the CACHE variables.

havogt replied:
@albestro I don't see the downsides here. Can you explain your exact use-case where this is a problem?

teonnik replied:
> To be honest, I'm not sure I understand the desired semantics. IIRC, this code will:
> 
>     * If the variable is in cache, don't do anything (e.g. from a previous run, or from a user-supplied `-Dvar=...`).
> 
>     * If the variable is defined but not in cache, don't change, but also don't make a cache variable — this would be where at the calling site, something in the CMakeLists.txt has set it.
> 
>     * If the variable is not defined at all, then make a cache variable with a reasonable default.
> 
> 
> Is this correct?

@havogt That is correct. That is exactly the behaviour that I was going for.

> I want to be able to set values from the script, so that in my package config script, when I do the find_dependency, I can force the FindMKL package to replicate exactly the same options as configured at build time. With this solution, values set from the script will be not considered if the user sets a cache variable with -D from the command line (as you are pointing out in your first point).

@albestro My understanding is that values from the script will take precedence over `-D` options since they are normal CMake variables as opposed to CACHE variables. For example

```
set(foo "foo")
message(STATUS ${foo})
```
built with `cmake -Dfoo="bar"` still prints "foo". The normal variable takes precedence.

albestro replied:
Yes, I was wrong. I got confused, I was not considering the right order of referencing a variable name: the order is `normal variable > cache variable`, so if there is any normal variable, it will always have priority no matter if a cache variable exists or not. And it is exactly what I want to achieve in my use-case.

About not exposing the value in the cache if a value is set from the script, it can be ok: it is true that the user cannot see what the final value is, but it is also true that he/she does not get confused by a cache variable that gets always set to the value decided in the script.
In order to have this last behavior with CMake version <3.15, where they added the [CACHE{variable}](https://cmake.org/cmake/help/v3.15/manual/cmake-language.7.html#variable-references), I can just think of workarounds that are not worth it.

I'm very satisfied with this work. Thanks everyone! I'm going to employ it in my script.

albestro commented on FindMKL.cmake:
```diff
@@ -0,0 +1,227 @@
+    if (NOT OPENMPI_FOUND)
+        include(${CMAKE_CURRENT_LIST_DIR}/check_for_openmpi.cmake)
+        check_for_openmpi()
+    endif()
```
I marked https://github.com/eth-cscs/cmake-recipes/pull/1#discussion_r331005997 as unresolved.

Do you have any update on this? Is there a reason for not removing the external `check_for_openmpi` script?

teonnik committed:
Replace custom MPI vendor detection

- Use MPI_DETERMINE_LIBRARY_VERSION to identify the MPI vendor instead
  of the `check_for_openmpi()` script.

- Fix duplicate link in options comment

albestro commented on FindMKL.cmake:
```diff
@@ -1,38 +1,113 @@
+# * There are many caveats with CACHE variables in CMake. Before version 
+#   3.12, both the `option()` command and `set(... CACHE ...)` would 
+#   override normal variables if cached equivalents don't exist or they 
+#   exisit but their type is not specified (e.g. command line arguments: 
```
typo: *exisit*

albestro commented on FindMKL.cmake:
```diff
@@ -0,0 +1,222 @@
+    __mkl_find_library(MKL_SCALAPACK_LIB mkl_scalapack_${_mkl_lp})
+
+    string(FIND "${MPI_CXX_LIBRARY_VERSION_STRING}" "Open MPI" OMPI_POS)
+    if(OMPI_POS STREQUAL "-1")  # MPICH
```
*Minor thing*
It seems that you follow this convention in the rest of the script (I like it too):
>"For internal implementation, it is a generally accepted convention that variables starting with underscore are for temporary use only."

So, I would fix also the name of `OMPI_POS`.

albestro commented on FindMKL.cmake:
```diff
@@ -0,0 +1,222 @@
+if (MPI_FOUND)
+    __mkl_find_library(MKL_SCALAPACK_LIB mkl_scalapack_${_mkl_lp})
+
+    string(FIND "${MPI_CXX_LIBRARY_VERSION_STRING}" "Open MPI" OMPI_POS)
```
You did a great job about this! But, it opens a bigger question, that we should address in general for the entire repo:

CMake 3.15 `MPI_CXX_LIBRARY_VERSION_STRING`
CMake 3.10 -> 3.14 `MPI_CXX_LIBRARY_VERSION`
CMake 3.10 *<NOT AVAILABLE>*

See disccusion in https://github.com/eth-cscs/cmake-recipes/issues/5

teonnik replied:
If we choose to not use variables provided by `FindMPI`, I can easily get the information that I need using `execute_process(COMMAND mpirun --version OUTPUT_VARIABLE MPIRUN_OUTPUT)`. That way we are not as dependent on the CMake version.

Alternatively, we can bundle (i.e. copy) the FindMPI module from CMake 3.15 into this repository and update it with CMake releases. After all, the CMake language barely changes between versions, we can use the new versions of the script with old versions of CMake itself.

albestro replied:
I don't know what others think about this, but what about something like this?

```cmake
cmake_minimum_required(VERSION REQUIRED 3.10)
# ...
if(CMAKE_VERSION VERSION_LESS 3.15)
  set(_MPI_CXX_IMPLEMENTATION ${MPI_CXX_LIBRARY_VERSION})
else()
  set(_MPI_CXX_IMPLEMENTATION ${MPI_CXX_LIBRARY_VERSION_STRING})
endif()
# ...
```

lukasm91 replied:
Yes this is what I would do (except `REQUIRED` is not needed)

teonnik replied:
For the moment, I decided to use the `execute_process(COMMAND mpirun --version OUTPUT_VARIABLE MPIRUN_OUTPUT)` approach. It's simple, shorter, faster and doesn't depend on CMAKE_VERSION. We can revise that in the future if people prefer other approaches.

albestro reviewed (commented):
Very good job! Nice implementation of the MPI implementation detection.

I would add Scalapack as COMPONENT (see specific comment).

lukasm91 commented on FindMKL.cmake:
```diff
@@ -0,0 +1,222 @@
+# Determine MKL's library folder
+#
+set(_mkl_libpath_suffix "lib/intel64")
+if(CMAKE_SIZEOF_VOID_P EQUAL 4) # 32 bit
```
You are not very consistent with spacings. I prefer having no spaces between command name and parentheses

lukasm91 commented on FindMKL.cmake:
```diff
@@ -0,0 +1,222 @@
+^^^^^^^^^^^^^
+
+  find_package(MKL REQUIRED OPTIONAL_COMPONENTS SCALAPACK)
+  target_link_libraries(... MKL::MKL)
```
What will our convention be for target names? I think, all uppercase is not very common.

lukasm91 commented on FindMKL.cmake:
```diff
@@ -0,0 +1,222 @@
+         __mkl_find_library(MKL_BLACS_LIB mkl_blacs_openmpi_${_mkl_lp})
+    endif()
+
+    find_package_handle_standard_args(MKL_SCALAPACK REQUIRED_VARS MKL_BLACS_LIB
```
as you have several places where you can fail, it might help to add `FAIL_MESSAGE`

teonnik replied:
Fail messages are generated by `find_package_handle_standard_args`. Where would you suggest I sprinkle `FAIL_MESSAGE` exactly?

lukasm91 commented on FindMKL.cmake:
```diff
@@ -0,0 +1,222 @@
+
+    set(_mkl_threading_backend "")
+    if(MKL_PARALLEL)
+        find_package(OpenMP REQUIRED)
```
`COMPOENTS CXX`?  and `find_dependency`

teonnik replied:
Hmm, I was considering using `find_dependency` but did not commit since CMake's documentation suggested that it's used for `Config` files.

lukasm91 replied:
Hmm, could question. I think you might be right, I am not sure about the reasoning, but I tend to agree

teonnik replied:
I grepped over all Find modules shipped with CMake, none of them use `find_dependency`. I suggest we keep using `find_package` instead.

lukasm91 commented on FindMKL.cmake:
```diff
@@ -0,0 +1,222 @@
+# Find ScaLAPACK
+#
+set(MPI_DETERMINE_LIBRARY_VERSION ON)
+find_package(MPI)
```
do you need all components? and you should use `find_dependency` from `CMakeFindDependencyMacro`

teonnik replied:
How does `find_dependency` work with components? MPI is only needed when the ScaLAPACK component is requested. A user may have `find_package(MKL REQUIRED OPTIONAL_COMPONENTS SCALAPACK)` in which case `REQUIRED` should not be propagated to `find_dependency(MPI)`.

lukasm91 replied:
I think find_dependency will not help you here, this is only about what should be done if a user already called `find_package(MPI)` before. These here seems trickier, `MKL_FIND_COMPONETS` might help, maybe?

lukasm91 reviewed (commented):
I quickly went through the file and added some comments. But looks good overall :+1:

havogt commented:
In general I think we don't need to review the cmake files to death. If we agree that a module goes in the right direction, we should take it and improve as we go.

teonnik committed:
Improve handling of MKL dependencies

- Search for ScaLAPACK component only when requested
- FindThreads, FindOpenMP and FindMPI are now properly handled as MKL
  dependencies.
- Fix typos and spacing issues
- Use lowercase for targets

albestro commented:
I'm ok with @havogt statement. Maybe it would be worth to open issues for changes that are not going to be delivered with this PR?

@havogt, @lukasm91 you should have the permissions to merge (in case not, let me know). Feel free to merge this as soon as you are ok with the review ;)
